File: android/vbt-app/app/src/main/java/com/vbt/app/data/ble/VbtBleManager.kt
package com.vbt.app.data.ble

import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCharacteristic
import android.content.Context
import android.util.Log
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import no.nordicsemi.android.ble.BleManager
import no.nordicsemi.android.ble.data.Data
import no.nordicsemi.android.ble.observer.ConnectionObserver
import java.nio.ByteBuffer
import java.nio.ByteOrder
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class VbtBleManager @Inject constructor(
    @ApplicationContext context: Context
) : BleManager(context) {

    companion object {
        private const val TAG = "VbtBleManager"
        private const val RECONNECT_ATTEMPTS = 5
        private const val RECONNECT_DELAY_MS = 1000L
    }

    private var liveVelocityChar: BluetoothGattCharacteristic? = null
    private var repResultChar: BluetoothGattCharacteristic? = null
    private var deviceStatusChar: BluetoothGattCharacteristic? = null
    private var commandChar: BluetoothGattCharacteristic? = null

    private val _liveVelocity = MutableStateFlow(0f)
    val liveVelocity: StateFlow<Float> = _liveVelocity

    private val _isLifting = MutableStateFlow(false)
    val isLifting: StateFlow<Boolean> = _isLifting

    private val _repCount = MutableStateFlow(0)
    val repCount: StateFlow<Int> = _repCount

    private val _repResult = MutableSharedFlow<RepFromDevice>(extraBufferCapacity = 10)
    val repResult: SharedFlow<RepFromDevice> = _repResult

    private val _connectionState = MutableStateFlow(BleConnectionState.DISCONNECTED)
    val connectionState: StateFlow<BleConnectionState> = _connectionState

    // Auto-reconnect: rozłączenie zainicjowane przez użytkownika (disconnectDevice())
    // nie powinno uruchamiać automatycznego wznawiania połączenia.
    private var userInitiatedDisconnect = false
    private var lastDevice: BluetoothDevice? = null

    private val managerScope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private var reconnectJob: Job? = null

    init {
        setConnectionObserver(object : ConnectionObserver {
            override fun onDeviceConnecting(device: BluetoothDevice) {}
            override fun onDeviceConnected(device: BluetoothDevice) {}
            override fun onDeviceFailedToConnect(device: BluetoothDevice, reason: Int) {}
            override fun onDeviceReady(device: BluetoothDevice) {
                reconnectJob?.cancel()
                _connectionState.value = BleConnectionState.CONNECTED
            }

            override fun onDeviceDisconnecting(device: BluetoothDevice) {}

            override fun onDeviceDisconnected(device: BluetoothDevice, reason: Int) {
                if (userInitiatedDisconnect) {
                    _connectionState.value = BleConnectionState.DISCONNECTED
                } else {
                    Log.w(TAG, "Niezamierzone rozłączenie BLE (reason=$reason) - próba wznowienia")
                    startAutoReconnect()
                }
            }
        })
    }

    override fun isRequiredServiceSupported(gatt: BluetoothGatt): Boolean {
        val service = gatt.getService(BleConstants.VBT_SERVICE_UUID) ?: return false
        liveVelocityChar = service.getCharacteristic(BleConstants.LIVE_VELOCITY_UUID)
        repResultChar = service.getCharacteristic(BleConstants.REP_RESULT_UUID)
        deviceStatusChar = service.getCharacteristic(BleConstants.DEVICE_STATUS_UUID)
        commandChar = service.getCharacteristic(BleConstants.COMMAND_UUID)
        return liveVelocityChar != null && repResultChar != null && commandChar != null
    }

    override fun initialize() {
        requestConnectionPriority(BluetoothGatt.CONNECTION_PRIORITY_HIGH).enqueue()
        requestMtu(64).enqueue()

        setNotificationCallback(liveVelocityChar).with { _, data ->
            parseLiveVelocity(data)
        }
        enableNotifications(liveVelocityChar).enqueue()

        setNotificationCallback(repResultChar).with { _, data ->
            parseRepResult(data)
        }
        enableNotifications(repResultChar).enqueue()
    }

    override fun onServicesInvalidated() {
        liveVelocityChar = null
        repResultChar = null
        deviceStatusChar = null
        commandChar = null
    }

    private fun parseLiveVelocity(data: Data) {
        if (data.size() < 6) return
        val bytes = data.value ?: return
        val buffer = ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN)

        val velocityRaw = buffer.short
        val flags = bytes[2].toInt() and 0xFF
        buffer.position(3)
        val repCountRaw = buffer.short.toInt() and 0xFFFF

        _liveVelocity.value = velocityRaw / 1000f
        _isLifting.value = (flags and 0x01) != 0
        _repCount.value = repCountRaw
    }

    private fun parseRepResult(data: Data) {
        val rep = RepPacketParser.parse(data.value) ?: return
        _repResult.tryEmit(rep)
    }

    fun sendCommand(command: ByteArray) {
        commandChar?.let {
            writeCharacteristic(it, command).enqueue()
        }
    }

    fun setExerciseParams(minLiftVel: Float, endLiftVel: Float, minRepDist: Float) {
        val data = byteArrayOf(
            BleConstants.CMD_SET_EXERCISE_PARAMS,
            (minLiftVel * 100).toInt().toByte(),
            (endLiftVel * 100).toInt().toByte(),
            (minRepDist * 100).toInt().toByte()
        )
        sendCommand(data)
    }

    fun connectToDevice(device: BluetoothDevice) {
        lastDevice = device
        userInitiatedDisconnect = false
        reconnectJob?.cancel()
        _connectionState.value = BleConnectionState.CONNECTING
        connect(device)
            .retry(3, 200)
            .useAutoConnect(true)
            .done {
                _connectionState.value = BleConnectionState.CONNECTED
            }
            .fail { _, status ->
                Log.w(TAG, "Połączenie BLE nieudane (status=$status)")
                _connectionState.value = BleConnectionState.DISCONNECTED
            }
            .enqueue()
    }

    fun reconnect() {
        lastDevice?.let { connectToDevice(it) }
    }

    private fun startAutoReconnect() {
        val device = lastDevice
        if (device == null) {
            _connectionState.value = BleConnectionState.DISCONNECTED
            return
        }
        _connectionState.value = BleConnectionState.RECONNECTING
        reconnectJob?.cancel()
        reconnectJob = managerScope.launch {
            delay(RECONNECT_DELAY_MS)
            if (userInitiatedDisconnect) return@launch
            connect(device)
                .retry(RECONNECT_ATTEMPTS, RECONNECT_DELAY_MS.toInt())
                .useAutoConnect(true)
                .done {
                    _connectionState.value = BleConnectionState.CONNECTED
                }
                .fail { _, status ->
                    Log.w(TAG, "Auto-reconnect BLE nieudany (status=$status)")
                    _connectionState.value = BleConnectionState.DISCONNECTED
                }
                .enqueue()
        }
    }

    fun disconnectDevice() {
        userInitiatedDisconnect = true
        reconnectJob?.cancel()
        _connectionState.value = BleConnectionState.DISCONNECTING
        disconnect()
            .done { _connectionState.value = BleConnectionState.DISCONNECTED }
            .fail { _, _ -> _connectionState.value = BleConnectionState.DISCONNECTED }
            .enqueue()
    }
}

File: ios/VBT/VBT/Core/BLE/VbtBleManager.swift
import CoreBluetooth
import Foundation
import Observation

/// Port `data/ble/VbtBleManager.kt` z Nordic Android BLE na CoreBluetooth.
///
/// NIEZWERYFIKOWANE NA SPRZĘCIE: ten plik był napisany bez dostępu do Xcode/fizycznego
/// iPhone'a. Zachowanie CoreBluetooth przy częstych notify (~co 0.1s z FF01) i przy
/// reconnект w tle wymaga testu na iPhone 15 Pro + realnym ESP32 przed dalszą rozbudową
/// (patrz IOS_PORT_PLAN.md, Faza 0 / punkt "spike ryzyka").
@Observable
final class VbtBleManager: NSObject {
    private static let reconnectAttempts = 5
    private static let reconnectDelay: TimeInterval = 1.0

    private(set) var liveVelocity: Float = 0
    private(set) var isLifting = false
    private(set) var repCount = 0
    private(set) var connectionState: BleConnectionState = .disconnected
    private(set) var discoveredPeripherals: [CBPeripheral] = []

    /// Konsumowane przez WorkoutViewModel — jedna wartość na powtórzenie odebrane z FF02.
    var onRepResult: ((RepFromDevice) -> Void)?

    private var central: CBCentralManager!
    private var peripheral: CBPeripheral?
    private var liveVelocityChar: CBCharacteristic?
    private var repResultChar: CBCharacteristic?
    private var deviceStatusChar: CBCharacteristic?
    private var commandChar: CBCharacteristic?

    // Auto-reconnect: rozłączenie zainicjowane przez użytkownika (disconnect())
    // nie powinno uruchamiać automatycznego wznawiania połączenia.
    private var userInitiatedDisconnect = false
    private var reconnectTask: Task<Void, Never>?

    override init() {
        super.init()
        central = CBCentralManager(delegate: self, queue: .main)
    }

    func startScan() {
        guard central.state == .poweredOn else { return }
        connectionState = .scanning
        discoveredPeripherals = []
        central.scanForPeripherals(withServices: [BleConstants.vbtServiceUUID])
    }

    func stopScan() {
        central.stopScan()
    }

    func connect(to device: CBPeripheral) {
        stopScan()
        userInitiatedDisconnect = false
        reconnectTask?.cancel()
        connectionState = .connecting
        peripheral = device
        device.delegate = self
        central.connect(device)
    }

    /// Ponowna próba połączenia z ostatnio używanym urządzeniem (np. przycisk "Reconnect"
    /// na WorkoutScreen po tym, jak auto-reconnect wyczerpał limit prób).
    func reconnect() {
        guard let peripheral else { return }
        connect(to: peripheral)
    }

    func disconnect() {
        userInitiatedDisconnect = true
        reconnectTask?.cancel()
        guard let peripheral else {
            connectionState = .disconnected
            return
        }
        connectionState = .disconnecting
        central.cancelPeripheralConnection(peripheral)
    }

    func sendCommand(_ bytes: [UInt8]) {
        guard let peripheral, let commandChar else { return }
        peripheral.writeValue(Data(bytes), for: commandChar, type: .withResponse)
    }

    func setExerciseParams(minLiftVel: Float, endLiftVel: Float, minRepDist: Float) {
        let bytes: [UInt8] = [
            BleConstants.Command.setExerciseParams.rawValue,
            UInt8(clamping: Int(minLiftVel * 100)),
            UInt8(clamping: Int(endLiftVel * 100)),
            UInt8(clamping: Int(minRepDist * 100))
        ]
        sendCommand(bytes)
    }

    private func startAutoReconnect() {
        guard let peripheral, !userInitiatedDisconnect else {
            connectionState = .disconnected
            return
        }
        connectionState = .reconnecting
        reconnectTask?.cancel()
        reconnectTask = Task { [weak self] in
            guard let self else { return }
            for attempt in 1...Self.reconnectAttempts {
                try? await Task.sleep(for: .seconds(Self.reconnectDelay))
                if Task.isCancelled || self.userInitiatedDisconnect { return }
                self.central.connect(peripheral)
                // Czekamy na wynik przez connectionState ustawiany w delegate; jeśli po
                // kolejnej próbie nadal nie jesteśmy połączeni, próbujemy dalej.
                try? await Task.sleep(for: .milliseconds(500))
                if self.connectionState == .connected { return }
                if attempt == Self.reconnectAttempts {
                    self.connectionState = .disconnected
                }
            }
        }
    }

    private func parseLiveVelocity(_ data: Data) {
        guard data.count >= 6 else { return }
        let velocityRaw = Int16(data[data.startIndex]) | (Int16(data[data.startIndex + 1]) << 8)
        let flags = data[data.startIndex + 2]
        let repCountRaw = UInt16(data[data.startIndex + 3]) | (UInt16(data[data.startIndex + 4]) << 8)

        liveVelocity = Float(velocityRaw) / 1000
        isLifting = (flags & 0x01) != 0
        repCount = Int(repCountRaw)
    }

    private func parseRepResult(_ data: Data) {
        guard let rep = RepPacketParser.parse(data) else { return }
        onRepResult?(rep)
    }
}

extension VbtBleManager: CBCentralManagerDelegate {
    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        // Nic nie robimy automatycznie — UI (ConnectScreen) decyduje kiedy zacząć skan,
        // pokazując stan `central.state` (np. poproszenie o włączenie Bluetootha).
    }

    func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String: Any], rssi RSSI: NSNumber) {
        guard !discoveredPeripherals.contains(where: { $0.identifier == peripheral.identifier }) else { return }
        discoveredPeripherals.append(peripheral)
    }

    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        peripheral.discoverServices([BleConstants.vbtServiceUUID])
    }

    func centralManager(_ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?) {
        connectionState = .disconnected
    }

    func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
        liveVelocityChar = nil
        repResultChar = nil
        deviceStatusChar = nil
        commandChar = nil

        if userInitiatedDisconnect {
            connectionState = .disconnected
        } else {
            startAutoReconnect()
        }
    }
}

extension VbtBleManager: CBPeripheralDelegate {
    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        guard let service = peripheral.services?.first(where: { $0.uuid == BleConstants.vbtServiceUUID }) else { return }
        peripheral.discoverCharacteristics(
            [BleConstants.liveVelocityUUID, BleConstants.repResultUUID, BleConstants.deviceStatusUUID, BleConstants.commandUUID],
            for: service
        )
    }

    func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {
        guard let characteristics = service.characteristics else { return }
        for characteristic in characteristics {
            switch characteristic.uuid {
            case BleConstants.liveVelocityUUID:
                liveVelocityChar = characteristic
                peripheral.setNotifyValue(true, for: characteristic)
            case BleConstants.repResultUUID:
                repResultChar = characteristic
                peripheral.setNotifyValue(true, for: characteristic)
            case BleConstants.deviceStatusUUID:
                deviceStatusChar = characteristic
            case BleConstants.commandUUID:
                commandChar = characteristic
            default:
                break
            }
        }

        // "Ready" = mamy wymagane charakterystyki (odpowiednik `isRequiredServiceSupported`
        // z Nordic BleManager, które warunkowało `onDeviceReady`).
        if liveVelocityChar != nil, repResultChar != nil, commandChar != nil {
            reconnectTask?.cancel()
            connectionState = .connected
        }
    }

    func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
        guard let data = characteristic.value else { return }
        switch characteristic.uuid {
        case BleConstants.liveVelocityUUID:
            parseLiveVelocity(data)
        case BleConstants.repResultUUID:
            parseRepResult(data)
        default:
            break
        }
    }
}
